matchers/parse/parse.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parse
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/alertmanager/pkg/labels"
+)
+
+var (
+	ErrEOF          = errors.New("end of input")
+	ErrNoOpenBrace  = errors.New("expected opening brace")
+	ErrNoCloseBrace = errors.New("expected close brace")
+	ErrNoLabelName  = errors.New("expected label name")
+	ErrNoLabelValue = errors.New("expected label value")
+	ErrNoOperator   = errors.New("expected an operator such as '=', '!=', '=~' or '!~'")
+	ErrExpectedEOF  = errors.New("expected end of input")
+)
+
+// Parser reads the sequence of tokens from the lexer and returns either a
+// series of matchers or an error. An error can occur if the lexer attempts
+// to scan text that does not match the expected grammar, or if the tokens
+// returned from the lexer cannot be parsed into a complete series of matchers.
+// For example, the input is missing an opening bracket, has missing label
+// names or label values, a trailing comma, or missing closing bracket.
+type Parser struct {
+	done         bool
+	err          error
+	hasOpenParen bool
+	input        string
+	lexer        Lexer
+	matchers     labels.Matchers
```

Comment:
```suggestion
        // To control the final state of the parser and aid idempotency.
	done         bool
	err             error
	matchers  labels.Matchers
	
	lexer        Lexer
        // hasOpenParen helps keep track if the matches are are wrapped in `{}`.
	hasOpenParen bool
```

I don't think we need to store the input if the lexer already has. The only usage of the input is to calculate the total length when printing error messages - perhaps we can store just that or ask the lexer.

Reply:
I can store the len, which I think will save 8B (string header is 16B), if that works? We can ask the lexer too, but need to change the signature of `accept` and `expect` to take `*Lexer`.

Reply:
The struct attributes are ordered alphabetical, will change them to what's recommended.